File: frontend/lib/api/liveAgent.ts
// ============================================================
// ThunderBots — Live Agent API Client (NEW)
// ============================================================
import { apiClient } from './client'

export type AgentStatus = 'online' | 'busy' | 'offline'
export type HandoffStatus = 'ai' | 'waiting' | 'active' | 'closed'

export interface Agent {
  id: string
  user_id: string
  name: string
  email: string
  status: AgentStatus
  active_chat_count: number
  max_concurrent_chats: number
  last_seen_at: string | null
}

export interface Handoff {
  id: string
  conversation_id: string
  workflow_id: string
  session_id: string
  status: HandoffStatus
  channel: string
  requested_by: string | null
  handoff_reason: string | null
  assigned_agent_id: string | null
  assigned_agent_name: string | null
  priority: number
  visitor_label: string
  last_message_preview: string | null
  last_message_at: string | null
  requested_at: string | null
  assigned_at: string | null
  closed_at: string | null
  created_at: string | null
  message_count: number | null
}

export interface HandoffMessage {
  id: string
  role: 'user' | 'bot' | 'agent' | 'system'
  content: string
  node_type: string | null
  created_at: string | null
}

export interface DashboardStats {
  active_chats: number
  waiting_chats: number
  closed_chats: number
  agents_online: number
  agents_busy: number
  agents_offline: number
}

export interface ConversationFilters {
  status?: HandoffStatus
  channel?: string
  agent_id?: string
  search?: string
  limit?: number
  offset?: number
  owner_id?: string
}

export const liveAgentApi = {
  getAgents: async (owner_id?: string): Promise<{ agents: Agent[] }> => {
    const { data } = await apiClient.get('/live-agent/agents', { params: { owner_id } })
    return data
  },

  updateMyStatus: async (status: AgentStatus, owner_id?: string, max_concurrent_chats?: number) => {
    const { data } = await apiClient.put('/live-agent/agents/me/status', { status, max_concurrent_chats }, {
      params: { owner_id },
    })
    return data
  },

  getDashboardStats: async (owner_id?: string): Promise<DashboardStats> => {
    const { data } = await apiClient.get('/live-agent/dashboard/stats', { params: { owner_id } })
    return data
  },

  getConversations: async (
    filters: ConversationFilters = {}
  ): Promise<{ items: Handoff[]; total: number; limit: number; offset: number }> => {
    const { owner_id, ...rest } = filters
    const { data } = await apiClient.get('/live-agent/conversations', { params: { owner_id, ...rest } })
    return data
  },

  getConversation: async (handoffId: string, owner_id?: string): Promise<{ handoff: Handoff; messages: HandoffMessage[] }> => {
    const { data } = await apiClient.get(`/live-agent/conversations/${handoffId}`, { params: { owner_id } })
    return data
  },

  takeOver: async (handoffId: string, owner_id?: string): Promise<Handoff> => {
    const { data } = await apiClient.post(`/live-agent/conversations/${handoffId}/take-over`, {}, { params: { owner_id } })
    return data
  },

  returnToAi: async (handoffId: string, owner_id?: string): Promise<Handoff> => {
    const { data } = await apiClient.post(`/live-agent/conversations/${handoffId}/return-to-ai`, {}, { params: { owner_id } })
    return data
  },

  close: async (handoffId: string, owner_id?: string): Promise<Handoff> => {
    const { data } = await apiClient.post(`/live-agent/conversations/${handoffId}/close`, {}, { params: { owner_id } })
    return data
  },

  sendMessage: async (handoffId: string, content: string, owner_id?: string): Promise<HandoffMessage> => {
    const { data } = await apiClient.post(`/live-agent/conversations/${handoffId}/messages`, { content }, { params: { owner_id } })
    return data
  },
}

File: backend/app/api/v1/live_agent.py
"""
ThunderBots Live Agent API (NEW)

Every route requires an authenticated User (get_current_user — same as the
rest of the app). Routes are scoped to a `owner_id` workspace: it defaults
to the caller's own user id (a solo bot owner acting as their own agent),
and also accepts an explicit `owner_id` query param for a *team* member
acting as an agent for another user's workspace — validated by reusing the
existing Team/TeamMember model (any role is enough to act as a live agent;
inviting/removing team members is unchanged, handled entirely by
api/v1/teams.py).

Does not touch Builder, Workflow Engine, Runtime, or AI Engine — reads/
writes only agent_profiles, live_agent_handoffs, and the pre-existing
conversations/messages tables.
"""
import logging
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.auth import get_current_user
from app.models.user import User
from app.models.team import Team, TeamMember
from app.services import live_agent_service as svc

router = APIRouter()
logger = logging.getLogger(__name__)


async def resolve_owner_scope(
    owner_id: Optional[str] = Query(None, description="Workspace owner id; defaults to the caller"),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> str:
    if not owner_id or owner_id == user.id:
        return user.id
    # Team member acting as a live agent for another user's workspace —
    # reuses the existing Team model; any membership role is sufficient.
    result = await db.execute(
        select(TeamMember).join(Team, Team.id == TeamMember.team_id)
        .where(Team.created_by == owner_id, TeamMember.user_id == user.id)
    )
    if not result.scalar_one_or_none():
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not a member of this workspace's team")
    return owner_id


class AgentStatusUpdate(BaseModel):
    status: str = Field(pattern="^(online|busy|offline)$")
    max_concurrent_chats: Optional[int] = Field(default=None, ge=1, le=50)


class TakeOverRequest(BaseModel):
    pass


class AgentMessageRequest(BaseModel):
    content: str = Field(min_length=1, max_length=8000)


class RequestHandoffBody(BaseModel):
    session_id: str
    workflow_id: str
    reason: Optional[str] = None
    channel: str = "web_chat"


# ── Agent status ─────────────────────────────────────────────────────────

@router.get("/agents")
async def get_agents(owner_id: str = Depends(resolve_owner_scope), db: AsyncSession = Depends(get_db)):
    return {"agents": await svc.list_agents(db, owner_id=owner_id)}


@router.put("/agents/me/status")
async def update_my_status(
    body: AgentStatusUpdate,
    owner_id: str = Depends(resolve_owner_scope),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    profile = await svc.set_agent_status(
        db, owner_id=owner_id, user_id=user.id, status=body.status,
        max_concurrent_chats=body.max_concurrent_chats,
    )
    return {
        "id": profile.id, "status": profile.status,
        "active_chat_count": profile.active_chat_count,
        "max_concurrent_chats": profile.max_concurrent_chats,
    }


# ── Dashboard ────────────────────────────────────────────────────────────

@router.get("/dashboard/stats")
async def get_dashboard_stats(owner_id: str = Depends(resolve_owner_scope), db: AsyncSession = Depends(get_db)):
    return await svc.dashboard_stats(db, owner_id=owner_id)


@router.get("/conversations")
async def get_conversations(
    status_filter: Optional[str] = Query(None, alias="status", pattern="^(ai|waiting|active|closed)$"),
    channel: Optional[str] = None,
    agent_id: Optional[str] = None,
    search: str = "",
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
    owner_id: str = Depends(resolve_owner_scope),
    db: AsyncSession = Depends(get_db),
):
    return await svc.list_conversations(
        db, owner_id=owner_id, status=status_filter, channel=channel, agent_id=agent_id,
        search=search, limit=limit, offset=offset,
    )


@router.get("/conversations/{handoff_id}")
async def get_conversation(
    handoff_id: str, owner_id: str = Depends(resolve_owner_scope), db: AsyncSession = Depends(get_db),
):
    try:
        return await svc.get_conversation_detail(db, handoff_id=handoff_id, owner_id=owner_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


# ── Handoff actions ──────────────────────────────────────────────────────

@router.post("/handoff/request")
async def request_handoff(
    body: RequestHandoffBody, owner_id: str = Depends(resolve_owner_scope),
):
    """Manual/API-triggered handoff request — the AI Agent node and the
    visitor-facing 'Talk to a human' button both ultimately call the same
    services.live_agent_service.request_handoff used here."""
    try:
        return await svc.request_handoff(
            session_id=body.session_id, workflow_id=body.workflow_id, owner_id=owner_id,
            channel=body.channel, reason=body.reason, requested_by="agent",
        )
    except Exception as e:
        logger.error(f"request_handoff failed: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail="Failed to request handoff")


@router.post("/conversations/{handoff_id}/take-over")
async def take_over_conversation(
    handoff_id: str, owner_id: str = Depends(resolve_owner_scope), user: User = Depends(get_current_user),
):
    try:
        return await svc.take_over(handoff_id=handoff_id, agent_user_id=user.id, owner_id=owner_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/conversations/{handoff_id}/return-to-ai")
async def return_conversation_to_ai(
    handoff_id: str, owner_id: str = Depends(resolve_owner_scope), user: User = Depends(get_current_user),
):
    try:
        return await svc.return_to_ai(handoff_id=handoff_id, owner_id=owner_id, actor_user_id=user.id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/conversations/{handoff_id}/close")
async def close_conversation_route(
    handoff_id: str, owner_id: str = Depends(resolve_owner_scope),
):
    try:
        return await svc.close_conversation(handoff_id=handoff_id, owner_id=owner_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/conversations/{handoff_id}/messages")
async def send_message(
    handoff_id: str, body: AgentMessageRequest,
    owner_id: str = Depends(resolve_owner_scope), user: User = Depends(get_current_user),
):
    try:
        return await svc.send_agent_message(
            handoff_id=handoff_id, owner_id=owner_id, agent_user_id=user.id, content=body.content,
        )
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
